optuna/samplers/base.py

```diff
@@ -1,28 +1,96 @@
 import abc
 import six
 
+import optuna
 from optuna.distributions import BaseDistribution  # NOQA
-from optuna.storages.base import BaseStorage  # NOQA
+from optuna.structs import FrozenTrial  # NOQA
+from optuna import types
+
+if types.TYPE_CHECKING:
+    from typing import Dict  # NOQA
+
+    from optuna.study import RunningStudy  # NOQA
 
 
 @six.add_metaclass(abc.ABCMeta)
 class BaseSampler(object):
     """Base class for samplers."""
 
-    @abc.abstractmethod
-    def sample(self, storage, study_id, param_name, param_distribution):
-        # type: (BaseStorage, int, str, BaseDistribution) -> float
+    def define_relative_search_space(self, study, trial):
+        # type: (RunningStudy, FrozenTrial) -> Dict[str, BaseDistribution]
+        """Define the search space used in the target trial.
+
+        The search space defined by this method will be used as an argument of
+        :func:`optuna.samplers.BaseSampler.sample_relative` method.
+
+        If a parameter that is not contained in the space is requested in an object function,
+        the value will be sampled by using :func:`optuna.samplers.BaseSampler.sample_independent`
+        method.
+
+        The default implementation returns the result of
+        :func:`optuna.study.RunningStudy.full_search_space` call.
+
+        Args:
+            study:
+                Target study object.
+            trial:
+                Target trial object.
+
+        Returns:
+            A dictionary containing the parameter names and parameter's distributions.
+
+        """
+
+        return study.full_search_space
+
+    def sample_relative(self, study, trial, search_space):
+        # type: (RunningStudy, FrozenTrial, Dict[str, BaseDistribution]) -> Dict[str, float]
+        """Sample parameters based on the previous trials and the given search space.
+
+        This method is called once just after each trial has started.
+
+        Note that this method is not supposed to be called by library users. Instead,
+        :class:`optuna.trial.Trial` provides user interfaces to sample parameters in an objective
+        function.
+
+        The default implementation of this method always returns :obj:`{}`.
+
+        Args:
+            study:
+                Target study object.
+            trial:
+                Target trial object.
+            param_name:
+                Name of the sampled parameter.
+            param_distribution:
+                Distribution object that specifies a prior and/or scale of the sampling algorithm.
+
+        Returns:
+            A float value in the internal representation of Optuna.
```

Comment:
This doesn't match with actual return value.

```suggestion
            A dictionary containing the parameter names and the values that are the internal representations of Optuna.
```